Remove uses of Expr.free_names in Un_cps

mshinwell opened the pull request:
This is the first stage in removing the free name caches from terms, which I suspect is causing performance problems.  (It's potentially possible to rectify it by only populating the caches when terms are being reconstructed after simplification, but the code to do this seems fragile.)

This needs a bit of tidying but review and testing would be appreciated, in particular to make sure that Un_cps is still doing the expected inlining (for let expressions, linearly-used continuation parameters and linearly-used results of applications).

The next step will be to remove the call to `Expr.free_names` from `Closure_conversion`, which I'll do by removing the denesting code, that I don't think we actually need.  After that, we should be in a position to propagate the free name set through `UA` in the simplifier, which I hope will be straightforward.  Then the new optional arguments to the various term creation functions can be provided from that set, rather than the caches.

Gbury commented on middle_end/flambda/to_cmm/un_cps.ml:
```diff
@@ -1321,7 +1364,9 @@ and let_dynamic_set_of_closures env res body closure_vars decls elts =
       let v = Var_in_binding_pos.var v in
       let e, effs = get_closure_by_offset env soc_cmm_var cid in
-      let_expr_bind body acc v e effs
+      (* It doesn't seem important to know the number of occurrences here. *)
```
Actually, it's one of the few places where it's reasonable to actually take into account the number of occurrences, in order to avoid creating unnecessary bindings (I'm not sure if such bindings could be eliminated by later passes). It's not crucial, but it would be nice to not bind closures that are not used at all (e.g. in cases where only one of the recursive functions is returned).

mshinwell replied:
This has been fixed

Gbury reviewed (commented):
The un_cps code looks good as far as I can tell. I'll try this and look at some examples tomorrow to verify inlining and other optimizations still work correctly (I wonder if we could have automated tests for this kind of things...)

mshinwell commented:
@Gbury I've changed various other things, so please could you check again - thanks.

Gbury commented on middle_end/flambda/to_cmm/un_cps_closure.ml:
```diff
@@ -382,7 +382,10 @@ module Greedy = struct
+          match (used_closure_vars : _ Or_unknown.t) with
+          | Unknown -> true
+          | Known used_closure_vars ->
+            Var_within_closure.Set.mem clos_var used_closure_vars)
```
This match may deserve to be factored out in some helper function `is_closure_var_used` somewhere

mshinwell replied:
Fixed

Gbury commented on middle_end/flambda/to_cmm/un_cps_env.mli:
```diff
@@ -49,7 +49,7 @@ val mk :
   Exported_code.t ->
   Continuation.t -> Continuation.t ->
-  Var_within_closure.Set.t -> t
+  used_closure_vars:Var_within_closure.Set.t Or_unknown.t -> t
```
The doc comment should be updated to reflect the fact that the last argument is now named.

mshinwell replied:
Fixed

Gbury reviewed (commented):
Had a quick look and it seems good. I'll check more attentively and with some example next week.

mshinwell commented:
@Gbury Could you please do the further checks you mentioned above?  I'll address the review comments soon.

mshinwell commented:
@Gbury I've rebased this and fixed the two things you mentioned.  Could you please do whatever further checks you wanted to do, so we can merge this?  Thanks

Gbury reviewed (approved):
Good to go, ^^
